[SYCL][Graph] Enable L0 optimizations (no profiling mode) (#358)
reble/llvm: sycl/doc/extensions/experimental/sycl_ext_oneapi_graph.asciidoc

  public:
    assume_buffer_outlives_graph() = default;
};

> 

Reviewer: missing `class enable_profiling {` with indent matching the above properties

graph. Destroying that buffer during the lifetime of a `command_graph`
constructed with this property results in undefined behavior.

===== Enable-Profiling Property [[enable-profiling]]

> The `property::graph::enable_profiling` property enable profiling
> the executable_graph.
> This property is passed to the `command_graph::finalize()` function.

Reviewer 2: ```suggestion
The `property::graph::enable_profiling` property can be passed to the 
`command_graph::finalize()` function and enables profiling support 
for the returned `command_graph<graph_state::executable>`.
```

If this property is not passed to the finalize function, certain optimizations,
which prevent graph profiling, may be applied to this executable graph.
In this case, an error will be thrown when attempting to profiling an event from 
a graph submission that was created without this property.

> Note that this limitation currently only applies to the zero-level backend. 

Reviewer: I think we should remove this line about level zero, and all backends should throw if this property isn't set for more consistent behaviour.

graph. Destroying that buffer during the lifetime of a `command_graph`
constructed with this property results in undefined behavior.

===== Enable-Profiling Property [[enable-profiling]]

The `property::graph::enable_profiling` property can be passed to the 
`command_graph::finalize()` function and enables profiling support 
for the returned `command_graph<graph_state::executable>`.
Passing this property to the finalize function implies disabling 
certain optimizations to enable graph profiling.
As a result, the execution time of a graph finalized with profiling enabled 
is longer than that of a graph without profiling capability.

> In this case, an error will be thrown when attempting to profiling an event 

Reviewer: ```suggestion
An error will be thrown when attempting to profile an event 
```

graph. Destroying that buffer during the lifetime of a `command_graph`
constructed with this property results in undefined behavior.

===== Disable-In-Order-Optimization Property [[disable-in-order-optimization]]

The `property::graph::disable_in_order_optimization` property disables 
an optimization which can be applied to in-order graph to 
reduce its execution time.
This property is passed to the `command_graph::finalize()` function.
Note that this property must be passed to the finalize function if users 

> want to profile the exec_graph execution.

Reviewer: This is quite a specific implementation detail to expose to the SYCL user, I think a better way to express the property would be `property::graph::enable_profiling` to align with queues, without going into the details of optimizations that will be disabled as there could be other optimizations in future that won't be compatible.

We should also specify an error is thrown when trying to profiling an event from a graph submission that was created without this property being set on the exec graph.

Author: I agree. A `property::graph::enable_profiling` could also cover other optimizations such as disabling profiling on events.

graph. Destroying that buffer during the lifetime of a `command_graph`
constructed with this property results in undefined behavior.

===== Enable-Profiling Property [[enable-profiling]]

The `property::graph::enable_profiling` property can be passed to the 
`command_graph::finalize()` function and enables profiling support 
for the returned `command_graph<graph_state::executable>`.
Passing this property to the finalize function implies disabling 
certain optimizations to enable graph profiling.
As a result, the execution time of a graph finalized with profiling enabled 
is longer than that of a graph without profiling capability.
An error will be thrown when attempting to profile an event 

> from a graph submission that was created without this property.

Reviewer 3: For Record & Replay, do we need to pass both the queue profiling property and the graph property to be able to profile graph events?

Author: Yes, we need both in the current implementation.

Reviewer 3: Hmm, I'm not super keen on requiring both for R&R, some thoughts:

- Perhaps the property would be better served being for nodes in the explicit API? That way there is a clear single mechanism for each API, while also allowing for more granularity (not forcing profiling on all nodes).
- I'm guessing both are required because otherwise the returned events from command recording would have profiling disabled, but after finalize they would then have profiling enabled and this is confusing to the user? However, I don't like that passing the queue property but not the graph one is basically a no-op, I think that is more confusing.

Author: I've updated the code (and the spec) accordingly.
Now, users must enable the profiling on at least one node to enable the profiling of the graph,
This can be done by using either the `property::graph::enable_profiling` property in building mode or the `property::queue::enable_profiling` on the recorded queue (Record&Replay API).
Note that the `property::queue::enable_profiling` must also set on the queue used to submit the graph. The event submission timestamp is indeed managed by the runtime, and is only set when submit queue has the flag. This might change with the profiling tag...

Author: It's not implemented yet, but I think it might be useful to add a new API to disable profiling on nodes. Such as `node::update_properties(const property_list &PropList)` with a new property `property::queue::disable_profiling`?

Parameters:

* `propList` - Optional parameter for passing properties. The only defined 

> property is `disable_in_order_optimization`. See <<disable_in_order_optimization, Disable-In-Order-Optimization>>

Reviewer: old property name
